### DISCUSSION TITLE
vcPolygon的性能问题

### DISCUSSION THREAD

**Discussion by MangMax** [zouyaoji/vue-cesium#347] [Q&A] [answered]
我使用vcPolygon加载3-5千个面出现明显卡顿，最高只有十几帧，但是使用polygonGeometry在VcPrimitive中加载卡顿缓解了许多，稳定在二三十帧，请问vcPolygon是否有办法优化效率或者我的使用姿势可能有问题？

**Comment by zouyaoji** [comment]
如果不是这样写的，这样试试，避免使用很多VcPolygon标签

<vc-collection-primitive :polygons="polygons"></vc-collection-primitive>

如果已经是这样，麻烦给一个复现例子，我空了测测。

**Reply by MangMax** [reply]
https://codepen.io/1061875478/pen/qBMBBaR
这是一个简单的例子，直接加载三千个Polygon，界面明显卡顿

**Reply by zouyaoji** [reply]
我这不动还是能上40帧，一动就变20多了，确实够呛~
![image](https://user-images.githubusercontent.com/24430116/218744075-a615a37f-b3f3-4e4f-aac0-46ae663f12f6.png)

我还没看到怎么优化，源码在这，欢迎共同探讨
https://github.com/zouyaoji/vue-cesium/blob/dev/packages/shared/src/PolygonPrimitive.ts

**Reply by MangMax** [reply]
目前确实还没找到优化办法，不确定是哪里出了问题。我尝试了mars3d的CombinePrimitive可以加载海量面并且不吃太高性能，但是没找到源码
![image](https://user-images.githubusercontent.com/68525017/218935804-8d4adfe9-b545-4427-8b7f-47f0abc09c15.png)
另外，PolygonPrimitive 的id是否应该改成any类型，目前是string，应该和下面这个保持一致
![image](https://user-images.githubusercontent.com/68525017/218936292-d2520528-3bba-44bd-85d9-8023a2b20f29.png)
否则不支持传入自定义的面属性

**Reply by MangMax** [reply]
https://www.cnblogs.com/mazhenyu/p/11534376.html
看看这篇文章是否有用，我目前使用这种方式创建的，性能会好一些，但还是达不到创建几万级别的优化
![image](https://user-images.githubusercontent.com/68525017/218938482-74b049f0-6001-4937-adb4-befce1ce3ddf.png)

**Reply by zouyaoji** [reply]
好的哈。我回头也学习一下。

mars3d 那个在线地址也发一下呗

**Reply by MangMax** [reply]
http://mars3d.cn/doc.html#map/graphic
http://mars3d.cn/editor-vue.html?id=graphic/combine/polygon

**Reply by zouyaoji** [reply]
我看了，他们这个确实很厉害，先mark了，等空了看看能不能偷师学艺到一些东西。

**Reply by zouyaoji** [reply]
cnblogs那个，也想请你帮忙试试看看放到vue-cesium中咋样，如果有效果，可以提个pr，把源码改了
https://github.com/zouyaoji/vue-cesium/blob/dev/packages/shared/src/PolygonPrimitive.ts

**Reply by MangMax** [reply]
好的，我有空再试一下

**Reply by MangMax** [reply]
最近做了一下测试，虽然性能有提升，不会明显掉帧，但是渲染速度还是感人，感觉还是达不到渲染海量数据的效果。已经转向搭建geoserver来解决了。

**Reply by zouyaoji** [reply]
geoserver可以的，缺点就是不能选中换颜色，或者是我不知道怎么弄？

**Reply by MangMax** [reply]
确实不行，性能牺牲灵活性

**Reply by zouyaoji** [reply]
选中的时候获取到那个面的数据，在前端单独加一下~ 如果面节点不多的话

**Reply by MangMax** [reply, marked as answer]
> 选中的时候获取到那个面的数据，在前端单独加一下~ 如果面节点不多的话

确实可以这样做，但是如果是半透的那颜色也会不达预期
